ch_backup/storage/engine/s3/s3_engine.py

```diff
@@ -88,15 +89,36 @@ def delete_files(self, remote_paths: Sequence[str]) -> None:
         Delete multiple files from S3
         """
         # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.delete_objects
-        if self._bulk_delete_enabled:
-            objects_to_delete: Sequence = [
+        if not self._bulk_delete_enabled:
+            for remote_path in remote_paths:
+                self.delete_file(remote_path)
+            return
+
+        try:
+            objects_to_delete: list = [
                 {"Key": path.lstrip("/")} for path in remote_paths
             ]
             self._s3_client.delete_objects(
                 Bucket=self._s3_bucket_name, Delete={"Objects": objects_to_delete}
             )
-        else:
-            for remote_path in remote_paths:
+        except ClientError as e:
+            if "MalformedXML" not in repr(e):
+                raise
+
+            objects_to_delete_single: list = []
+            objects_to_delete_bulk: list = []
+            for path in remote_paths:
+                if contains_unusual_characters(path):
+                    objects_to_delete_single.append(path)
+                else:
+                    objects_to_delete_bulk.append({"Key": path.lstrip("/")})
+
+            if objects_to_delete_bulk:
+                self._s3_client.delete_objects(
```

Comment:
I think it is not related, but...:
 From doc: [The request can contain a list of up to 1000 keys that you want to delete.](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/delete_objects.html)
Can we limit size of the objects list in the single `delete_objects` call?

Reply:
Should be already handled by `DeleteMultipleStorageStage` from https://github.com/yandex/ch-backup/blob/master/ch_backup/storage/stages/storage.py